### PR TITLE
Fix CropMirrorNormalize crop_pos_x/y argument for the CPU

### PR DESCRIPTION
**Pull request by JanuszL**
- makes CropMirrorNormalize take into account crop_pos_x and
  crop_pos_y arguments

Signed-off-by: Janusz Lisiecki <jlisiecki@nvidia.com>

### CODE REVIEW

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 1]
```diff
@@ -118,9 +118,31 @@ void CropMirrorNormalize<CPUBackend>::RunHelper(SampleWorkspace *ws,
...
+
+  float crop_x_image_coord = spec_.GetArgument<float>("crop_pos_x", ws, ws->data_idx());
+  float crop_y_image_coord = spec_.GetArgument<float>("crop_pos_y", ws, ws->data_idx());
+
+  DALI_ENFORCE(crop_x_image_coord >= 0.f && crop_x_image_coord <= 1.f,
+      "Crop coordinates need to be in range [0.0, 1.0]");
+  DALI_ENFORCE(crop_y_image_coord >= 0.f && crop_y_image_coord <= 1.f,
+      "Crop coordinates need to be in range [0.0, 1.0]");
+
+  int crop_y = crop_y_image_coord * (H - crop_h_);
```
Comment (mzient) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]: The distribution is skewed. Should be:
```suggestion
  int crop_y = std::floor(crop_y_image_coord * (H - crop_h_) + 0.5f);
```

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc [hunk 2]
```diff
@@ -118,9 +118,31 @@ void CropMirrorNormalize<CPUBackend>::RunHelper(SampleWorkspace *ws,
...
+  int crop_x = crop_x_image_coord * (W - crop_w_);
```
Comment (mzient) [inline dali/pipeline/operators/fused/crop_mirror_normalize.cc]:
```suggestion
  int crop_x = std::floor(crop_x_image_coord * (W - crop_w_) + 0.5f);
```

Comment (JanuszL) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Will do (for GPU and crop itself as well).

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 1 above)
Comment (JanuszL) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Will do (for GPU and crop itself as well).

File: dali/test/python/test_pipeline.py [hunk 3]
```diff
@@ -324,6 +307,66 @@ def iter_setup(self):
...
+
+    class HybridPipe(Pipeline):
+        def __init__(self, batch_size, num_threads=1, device_id=0, num_gpus=1):
+            super(HybridPipe, self).__init__(batch_size,
+                                             num_threads,
+                                             device_id)
+            self.input = ops.CaffeReader(path = caffe_db_folder, shard_id = device_id, num_shards = num_gpus)
+            self.decode = ops.HostDecoder(device = "cpu", output_type = types.RGB)
+            self.cmnp_cpu = ops.CropMirrorNormalize(device = "cpu",
+                                                output_dtype = types.UINT8,
```
Comment (jantonguirao) [inline dali/test/python/test_pipeline.py]: indent

File: dali/test/python/test_pipeline.py [hunk 4]
```diff
@@ -324,6 +307,66 @@ def iter_setup(self):
...
+                                                output_layout = types.NHWC,
+                                                crop = (224, 224),
+                                                image_type = types.RGB,
+                                                mean = [0., 0., 0.],
+                                                std = [1., 1., 1.])
+
+            self.cmnp_gpu = ops.CropMirrorNormalize(device = "gpu",
+                                                output_dtype = types.UINT8,
```
Comment (jantonguirao) [inline dali/test/python/test_pipeline.py]: same here

File: dali/test/python/test_pipeline.py [hunk 5]
```diff
@@ -324,6 +307,66 @@ def iter_setup(self):
...
+                                                output_layout = types.NHWC,
+                                                crop = (224, 224),
+                                                image_type = types.RGB,
+                                                mean = [0., 0., 0.],
+                                                std = [1., 1., 1.])
+
+            self.crop_cpu = ops.Crop(device = "cpu", crop = (224, 224),
+                                                image_type = types.RGB,)
```
Comment (jantonguirao) [inline dali/test/python/test_pipeline.py]: and here

File: dali/test/python/test_pipeline.py [hunk 6]
```diff
@@ -324,6 +307,66 @@ def iter_setup(self):
...
             check_batch(outs[1], outs[3], batch_size)
 
+def test_cropmirrornormalize_cpu_vs_gpu():
+    def to_array(dali_out):
+        if isinstance(dali_out, TensorListGPU):
+            dali_out = dali_out.as_cpu()
+
+        return np.squeeze(dali_out.as_array())
+
+    def compare(val_1, val_2, reference=None):
```
Comment (jantonguirao) [inline dali/test/python/test_pipeline.py]: there are couple of functions to compare batches at the top of the file. I'd rather use those

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 2 above)
Comment (JanuszL) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Done

File: dali/pipeline/operators/fused/crop_mirror_normalize.cc (hunk 1 above)
Comment (JanuszL) [reply dali/pipeline/operators/fused/crop_mirror_normalize.cc]: Done

File: dali/test/python/test_pipeline.py (hunk 3 above)
Comment (JanuszL) [reply dali/test/python/test_pipeline.py]: Done

File: dali/test/python/test_pipeline.py (hunk 5 above)
Comment (JanuszL) [reply dali/test/python/test_pipeline.py]: Done

File: dali/test/python/test_pipeline.py (hunk 6 above)
Comment (JanuszL) [reply dali/test/python/test_pipeline.py]: Done

File: dali/test/dali_test_resize.h [hunk 7]
```diff
@@ -133,8 +134,8 @@ class GenericResizeTest : public DALISingleOpTest<ImgType> {
       if (resizeOptions & t_cropping) {
         finalImg = &crop_img;
 
-        const int crop_y = (rsz_h - crop_h) / 2;
-        const int crop_x = (rsz_w - crop_w) / 2;
+        const int crop_y = std::floor(0.5f * (rsz_h - crop_h) + 0.5f);
```
[the remark this answers is not in the document]
Comment (JanuszL) [reply dali/test/dali_test_resize.h]: Done

**Pull request merged**